### PR TITLE
Refine net time sync code for OneSync

### PR DESCRIPTION
**Pull request by Disquse**
### Goal of this PR

The goal of this PR is to improve the readability of `rage::netTimeSync` related code. Additionally, it aims to improve the maintainability of `CloneExperiments.cpp` by logically splitting certain parts of the file into appropriate chunks. The code refine isn't intended to be a full fledged refactor but rather a first step towards facilitating future refactoring efforts.

I invite everyone to participate in reviewing and suggesting improvements. Please remember that on the current stage I'm not planning to make it more complex than it is right now (i.e. adding abstraction layers, reworking the inner or depended code, etc).

### How is this PR achieving the goal

By moving the related code from `CloneExperiments.cpp` to separate `.cpp` and `.h` files, and making some minor tweaks.

### This PR applies to the following area(s)

FiveM, RedM

### Successfully tested on

**Game builds:**
FiveM: 1604, 2060, 2189, 2372. 2545, 2612, 2699, 2802, 2944, 3095
RedM: 1311, 1355, 1436, 1491

**Platforms:** Windows

### Checklist

- [x] Code compiles and has been tested successfully.
- [x] Code explains itself well and/or is documented.
- [x] My commit message explains what the changes do and what they are for.
- [x] No extra compilation warnings are added by these changes.

### Fixes issues

\-

### CODE REVIEW

**Comment by gottfriedleibniz** [comment]
Does our style guideline include rules on names/ordering/grouping of includes? How [netTimeSync](https://github.com/citizenfx/fivem/pull/2381/files#diff-50b327204f68ea36d25e6e6a6d13eddea56d85a2aaa32283126529f50daf07e8R1-R10) is currently structured is a bit unappealing to the eye.

File: code/components/gta-net-five/src/CloneExperiments.cpp [hunk 1]
```diff
@@ -2392,6 +2392,8 @@ static HookFunction hookFunction([]()
 		// memset in CNetworkDamageTracker::ctor
 		hook::put<uint32_t>(hook::get_pattern("48 89 11 48 8B D9 41 B8 80 00 00 00", 8), sizeof(float) * 256);
 	}
+
+	g_CNetPlayerOffset_PlayerInfo = *hook::get_pattern<int>("48 8B 81 ? 00 00 00 48 83 C0 20 C3", 3);;
```
Comment (gottfriedleibniz) [inline code/components/gta-net-five/src/CloneExperiments.cpp:2396]: ;;

File: code/components/gta-net-five/src/netTimeSync.cpp [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+	if (!icgi->OneSyncEnabled)
+	{
+		return g_origInitializeTime(timeSync, connectionMgr, flags, trustHost, sessionSeed, deltaStart, packetFlags, initialBackoff, maxBackoff);
+	}
+
+	timeSync->SetConnectionManager(connectionMgr);
+
+	return true;
+}
+template<int Build>
```
Comment (gottfriedleibniz) [inline code/components/gta-net-five/src/netTimeSync.cpp:45]: Newline

File: code/components/gta-net-five/src/netTimeSync.cpp [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+	if (xbr::IsGameBuildOrGreater<1355>())
+	{
+		return !(*g_netTimeSync<1355>)->IsInitialized();
+	}
+
+	return !(*g_netTimeSync<1311>)->IsInitialized();
+#endif
+}
+
+inline void TimeSyncMainGameFrameUpdate()
```
Comment (gottfriedleibniz) [inline code/components/gta-net-five/src/netTimeSync.cpp:203]: static / internal linkage?

**Comment by gottfriedleibniz** [comment]
One benefit here is that as we start refining/decoupling things - we'll figure out a nice convention that allows us to structure functionality a bit better (e.g., gamecode structure vs. hooking vs. onesync protocols, etc).

File: code/components/gta-net-five/src/netTimeSync.cpp [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+		(*g_netTimeSync<1311>)->HandleTimeSync(buf);
+	}
+#endif
+}
+
+static InitFunction initFunction([]()
+{
+	NetLibrary::OnNetLibraryCreate.Connect([](NetLibrary* lib)
+	{
+		lib->AddReliableHandler("msgTimeSync", [](const char* data, size_t len)
```
Comment (gottfriedleibniz) [inline code/components/gta-net-five/src/netTimeSync.cpp:266]: What is your opinion on moving the `xbr::IsGameBuildOrGreater` checks outside of the `ReliableHandlerType` callback. For example:

```cpp
// Beneath SetConnectionManager?
template<int Build>
static void HandleTimeSyncUpdatePacket(const char* data, size_t len)
{
	net::Buffer buf(reinterpret_cast<const uint8_t*>(data), len);
	(*g_netTimeSync<Build>)->HandleTimeSync(buf);
}
```

Then having the `InitFunction` something like (with brackets?):

```cpp
static InitFunction initFunction([]()
{
	static constexpr char kMsgTimeSync[] = "msgTimeSync"; // Eventually moved?
	NetLibrary::OnNetLibraryCreate.Connect([](NetLibrary* lib)
	{
#ifdef GTA_FIVE
		if (xbr::IsGameBuildOrGreater<2372>())
			lib->AddReliableHandler(kMsgTimeSync, HandleTimeSyncUpdatePacket<2372>);
		else if (xbr::IsGameBuildOrGreater<2060>())
			lib->AddReliableHandler(kMsgTimeSync, HandleTimeSyncUpdatePacket<2060>);
		else
			lib->AddReliableHandler(kMsgTimeSync, HandleTimeSyncUpdatePacket<1604>);
#elif IS_RDR3
		if (xbr::IsGameBuildOrGreater<1355>())
			lib->AddReliableHandler(kMsgTimeSync, HandleTimeSyncUpdatePacket<1355>);
		else
			lib->AddReliableHandler(kMsgTimeSync, HandleTimeSyncUpdatePacket<1311>);
#endif
	});
});
```

Comment (Disquse) [reply code/components/gta-net-five/src/netTimeSync.cpp:266]: Is there any reason to do it now? This approach as well as the one I'm currently using isn't something that we should keep anyway. The initializer lists makes more sense but would require more planning.

Comment (gottfriedleibniz) [reply code/components/gta-net-five/src/netTimeSync.cpp:266]: Agreed.

**Review by gottfriedleibniz — APPROVED** [review]

**Pull request merged**